app/frontend/wallet/shelley-wallet.ts

```diff
@@ -420,12 +428,116 @@ const ShelleyWallet = ({
+    validStakepools
+  ): Promise<DelegetionHistoryObject[]> {
+    const {delegations, rewards, withdrawals} = await blockchainExplorer.getDelegationHistory(
+      shelleyAccountInfo.accountPubkeyHex
+    )
+
+    const extractUrl = (poolHash) => {
+      const stakePool = validStakepools[poolHash]
+      if (stakePool) {
+        return stakePool.url
+      } else {
+        return null
+      }
+    }
+
+    const poolMetaUrls = distinct([
+      ...distinct(delegations.map((delegation) => extractUrl(delegation.poolHash))),
+      ...distinct(rewards.map((reward) => extractUrl(reward.poolHash))),
+    ]).filter((x) => x != null)
+
+    // Run requests for meta data in parallel
+    const metaDataPromises = poolMetaUrls.map((url: string) => ({
+      url,
+      metaDataPromise: getPoolInfo(url),
+    }))
+    const metaUrlToPoolNameMap = {}
+    for (const promise of metaDataPromises) {
+      const metaData = await promise.metaDataPromise
+      metaUrlToPoolNameMap[promise.url] = metaData.name
+    }
+
+    const poolHashToPoolName = (poolHash) => {
+      const poolName = metaUrlToPoolNameMap[extractUrl(poolHash)]
+      if (poolName) {
+        return poolName
+      } else {
+        return UNKNOWN_POOL_NAME
+      }
+    }
+
+    const parseStakePool = (delegationHistoryObject) => {
+      const stakePool: StakePool = {
+        id: delegationHistoryObject.poolHash,
+        name: poolHashToPoolName(delegationHistoryObject.poolHash),
+      }
+
+      return stakePool
+    }
+
+    // Prepare delegations
+    let oldPool: StakePool = null
+    const parsedDelegations = delegations
+      .map((delegation) => ({...delegation, time: new Date(delegation.time)}))
+      .sort((a, b) => a.time.getTime() - b.time.getTime()) // sort by time, oldest first
+      .map((delegation) => {
+        const stakePool: StakePool = parseStakePool(delegation)
+        const stakeDelegation: StakeDelegation = {
+          type: DelegetionHistoryItemType.StakeDelegation,
+          epoch: delegation.epochNo,
+          dateTime: new Date(delegation.time),
+          newStakePool: stakePool,
+          oldStakePool: oldPool,
+        }
+        oldPool = stakePool
+
+        return stakeDelegation
+      })
+      .reverse() // newest first
+
+    // Prepare rewards
+    const parsedRewards = rewards.map((reward) => {
+      const stakingReward: StakingReward = {
+        type: DelegetionHistoryItemType.StakingReward,
+        epoch: reward.epochNo,
+        dateTime: new Date(reward.time),
+        reward: reward.amount,
+        stakePool: parseStakePool(reward),
+      }
+
+      return stakingReward
+    })
+
+    // Prepare withdrawals
+    const parsedWithdrawals = withdrawals.map((withdrawal) => {
+      const rewardWithdrawal: RewardWithdrawal = {
+        type: DelegetionHistoryItemType.RewardWithdrawal,
+        epoch: withdrawal.epochNo,
+        dateTime: new Date(withdrawal.time),
+        credit: withdrawal.amount,
+      }
```

Comment:
I think we could do this in blockchain-explorer. Sorry for not being clear the first time but I think the blockchain-explorer should be responsible for getting the data from backend and returning them in the right format, and wallet should be responsible for doing something with that data. I think can do all of this "parsing" for delegations, rewards and withdrawals in blockchain-explorer and would rather pass the validPools back to the blockchain-explorer and let him do all of this logic (including getting the pool names). Once again sorry for not being 100 clear in the first case, and changing my mind in the second one but it makes a lot more sense to me this way.

Reply:
Done